src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionBase.cs

```diff
@@ -0,0 +1,202 @@
+                ? VSCommanding.CommandState.Available
+                : VSCommanding.CommandState.Unspecified;
+        }
+
+        protected static void InsertText(ArrayBuilder<TextChange> textChanges, int position, string text)
+        {
+            textChanges.Add(new TextChange(new TextSpan(position, 0), text));
+        }
+
+        protected static void DeleteText(ArrayBuilder<TextChange> textChanges, TextSpan span)
+        {
+            textChanges.Add(new TextChange(span, string.Empty));
+        }
+
+        internal bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, TCommand command, CommandExecutionContext context)
+        {
+            var title = GetTitle(command);
+            var message = GetMessage(command);
+
+            using (context.OperationContext.AddScope(allowCancellation: true, message))
+            {
+                var cancellationToken = context.OperationContext.UserCancellationToken;
+
+                var selectedSpans = textView.Selection.GetSnapshotSpansOnBuffer(subjectBuffer);
+                if (selectedSpans.IsEmpty())
+                {
+                    return true;
+                }
+
+                var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return true;
+                }
+
+                var service = document.GetLanguageService<ICommentSelectionService>();
+                if (service == null)
+                {
+                    return true;
+                }
+
+                var edits = CollectEdits(document, service, subjectBuffer, selectedSpans, command, cancellationToken).WaitAndGetResult(cancellationToken);
+
+                ApplyEdits(document, textView, subjectBuffer, service, title, edits);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Applies the requested edits and sets the selection.
+        /// This operation is not cancellable.
+        /// </summary>
+        private void ApplyEdits(Document document, ITextView textView, ITextBuffer subjectBuffer,
+            ICommentSelectionService service, string title, CommentSelectionResult edits)
+        {
+            // Create tracking spans to track the text changes.
+            var currentSnapshot = subjectBuffer.CurrentSnapshot;
+            var trackingSpans = edits.TrackingSpans
+                .SelectAsArray(textSpan => (originalSpan: textSpan, trackingSpan: CreateTrackingSpan(edits.ResultOperation, currentSnapshot, textSpan.TrackingTextSpan)));
+
+            // Apply the text changes.
+            using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+            {
+                document.Project.Solution.Workspace.ApplyTextChanges(document.Id, edits.TextChanges.Distinct(), CancellationToken.None);
```

Comment:
why do we need distinct? is textchanges supposed to be in order?

Reply:
so, I looked the code. so, it require no overlapping changes. I guess that is why you do Distinct? and implementation of both compiler and IDE seems takes care of unordered text changes, so that seems good.

Reply:
Yeah, unordered changes are taken care of, but there are overlapping cases (for intersecting block comments) that are most easily taken care of by distinct

Reply:
But I should mention changes should already be ordered as the selections are ordered.